### DISCUSSION TITLE
awkward results with DISCRETE_ADJOINT in the Quick Start tutorial

### DISCUSSION THREAD

**Discussion by frx-wintermute** [su2code/SU2#1994] [Q&A]
Hi!

I tried the [Quick Start] tutorial.

[Quick Start]: <https://su2code.github.io/docs_v7/Quick-Start/>

I built SU2 from source (from commit 3f68501a7ccb107f05a44c2aa5747797541fb683 , branch 'develop').
I took files 'mesh_NACA0012_inv.su2' and 'inv_NACA0012.cfg' from the same commit.

I managed to obtain a correct solution for the DIRECT problem (pressure field visualized with ParaView from 'flow.vtu' looks reasonably similar to the image shown in the tutorial).
Afterward, I changed the parameter MATH_PROBLEM from DIRECT to CONTINUOUS_ADJOINT in the configuration file and:

    $ ln -s restart_flow.dat solution_flow.dat
    $ SU2_CFD inv_adj_NACA0012.cfg

I obtained a correct adjoint solution (density adjoint field visualized with ParaView from 'adjoint.vtu' looks reasonably similar to the image shown in the tutorial).

Finally, I wanted to also try the algorithmic differentiation, so I changed the parameter MATH_PROBLEM to DISCRETE_ADJOINT in the configuration file and:

    $ SU2_CFD_AD inv_dadj_NACA0012.cfg

I failed to obtain a correct adjoint solution (I expected to see something really similar to what I got with the CONTINUOUS_ADJOINT method): the density adjoint field seems to have a completely different scale (instead of having values in the approximate range -200 ÷ +200, it has values in the range -0.264 ÷ +0.217 and the values not too close to zero are concentrated in an area near the farfield boundary, far from the airfoil).

What's wrong?
Does anything else need to be changed in the configuration file?

**Bug report checklist**

- [X] Consulting the build instructions (https://su2code.github.io/docs_v7/Build-SU2-Linux-MacOS/).

From the above mentioned commit, I built with:

    $ ./preconfigure.py --no-medi --no-opdi --no-mpp --no-coolprop
    $ export CXXFLAGS='-march=native -mtune=native -O3 -msse -mfpmath=sse'
    $ meson setup -Dwith-mpi=disabled -Denable-cgns=true -Denable-tecio=false -Denable-autodiff=true -Denable-directdiff=true --prefix=${HOME}/local build
    $ ninja -C build install

N.B.: CPU is an Intel Core i5-6600 CPU @ 3.30GHz

- [X] Looking for similar problems on GitHub or CFD-Online (https://www.cfd-online.com/Forums/su2/).

I haven't found any other GitHub issue about the Quick Start tutorial.

- [X] Updating to the newest version of SU2 (either develop, master, or the pre-built executables https://su2code.github.io/download.html).

I used the tip of the develop branch, as of yesterday.

**Desktop:**
 - OS: Debian GNU/Linux 12.0 'bookworm' (up-to-date Debian testing)
 - C++ compiler and version: g++ (Debian 12.2.0-14) 12.2.0
 - MPI implementation and version: irrelevant (built without MPI support)
 - SU2 Version: commit 3f68501a7ccb107f05a44c2aa5747797541fb683 , branch 'develop'

**Comment by pcarruscag** [comment]
The adjoint values from continuous and discrete approaches are not comparable

**Reply by frx-wintermute** [reply]
Thanks for your prompt reply.

> The adjoint values from continuous and discrete approaches are not comparable

This was not too clear from the tutorial text. Maybe some brief warning about this fact should be added to the tutorial text, so that people following the tutorial do not expect to see similar results from the two adjoint approaches...

Moreover, I think that an image showing the adjoint density field, as obtained with the discrete adjoint method, could be added to the tutorial, so that people have a chance to check that they are getting reasonable results with SU2_CFD_AD...

**Reply by frx-wintermute** [reply]
Do you agree that the tutorial should be enhanced?

**Reply by pcarruscag** [reply]
Yes that sounds good, what should be compared are the sensitivities. If you have time please go for it.

**Reply by frx-wintermute** [reply]
I would love to, but I have two issues with that.

First of all, I have familiarity with CFD, but I have never studied the adjoint method (I was hoping to learn more, while reading SU2 documentation, but right now I am stuck at the Quick Start tutorial...): I would rather avoid describing adjoint method results in the tutorial text, when I don't know what I am talking about!
And actually, I compared the sensititivies between the continuous and discrete approaches: they do not look similar. Hence, I am still convinced that something went wrong in my discrete adjoint method run.

And this leads us to the second issue: I am an SU2 beginner and I was following the Quick Start tutorial exactly to check that I managed to correctly build SU2 from source and to correctly run it. If I "enhance" the tutorial text by contributing wrong results, everybody else will read a wrong tutorial! I think that the tutorial should be enhanced by someone more knowledgeable than me about SU2...

**Reply by pcarruscag** [reply]
Can you post some pictures of what you are comparing?
There are very few volunteers that reply to this type of question, so please don't make this more time-consuming than it has to be.

**Reply by frx-wintermute** [reply]
After you told me that the adjoint variables are not comparable between the two approaches, I tried to compare the "Sensitivity" field, but the values seem to be basically zero for (almost) the entire domain. The only quantity that looks more or less comparable is the "Surface_Sensitivity". The two approaches give me the following two visualizations. I don't know whether you consider the two surface sensitivity fields similar enough to conclude that my runs are OK... Please let me know what you think.

![adj_surface_sens_field](https://user-images.githubusercontent.com/126598275/231456999-55d9a7bd-13ef-4803-ab71-63a8702104e0.png)
![dadj_surface_sens_field](https://user-images.githubusercontent.com/126598275/231457057-b27152db-81d5-450d-bf92-2a0e2da153ad.png)

**Reply by pcarruscag** [reply]
The continuous adjoint solver computes surface sensitivities directly (the volume sensitivities are 0).
The discrete adjoint solver computes mesh sensitivities (surface + volume), the surface is much more important.
SU2_DOT_AD converts the mesh sensitivities into surface sensitivities by accounting for the dependence of volume deformation on the surface deformation.
Then, these "total" surface sensitivities are used to compute the derivatives of the objective function with respect to the design variables (which define the surface deformation).
Therefore, surface sensitivities should be comparable between the 2 approaches, and the derivatives should also be close.
Close to sharp geometry features (e.g. trailing edges) is where continuous and discrete solvers tend to differ more.

**Reply by frx-wintermute** [reply]
Do you mean that an additional step is needed with SU2_DOT_AD? Or is this taken care of transparently behind the scenes, when SU2_CFD_AD is run?

I tried to compare the surface sensitivities on the "airfoil" boundary alone, by extracting comma-separated-value files (from ParaView) and then plotting the data together. The result seems to show some agreement, except for points close to sharp geometry features, as you said. Does that look OK?

[surf_sens_CMP_adj-dadj.pdf](https://github.com/su2code/SU2/files/11234090/surf_sens_CMP_adj-dadj.pdf)

**Reply by pcarruscag** [reply]
SU2_DOT_AD should change the surface sensitivity of the discrete adjoint a little, and if you want derivatives of design variables you need to run it, it is not part of SU2_CFD_AD.
The comparison looks ok.

**Reply by frx-wintermute** [reply]
Why is this additional step not described in the [Quick Start](https://su2code.github.io/docs_v7/Quick-Start/) tutorial? Could you please provide some brief instructions on how to perform this additional step?

Thanks a lot for your kind support!

**Comment by frx-wintermute** [comment]
I tried to run SU2_DOT_AD by using the same configuration file that was used with SU2_CFD_AD:

    $ ln -s restart_adj_cd.dat solution_adj_cd.dat
    $ SU2_DOT_AD inv_dadj_NACA0012.cfg

Apparently, this is not exactly the way to go.
I obtained three output files:

  * 'volume_sens.vtu', which seems to include sensitivities close to zero in almost all the domain
  * 'surface_sens.vtu', which includes a surface sensitivity only slightly changed with respect to the one produced by SU2_CFD_AD (as you said, see [surf_sens_SU2_DOT_AD_effect.pdf](https://github.com/su2code/SU2/files/11359758/surf_sens_SU2_DOT_AD_effect.pdf)), but also multiplied by -1 (I don't understand why)
  * 'of_grad.dat', which only includes one numerical value.

The content of the latter file is:

    $ cat of_grad.dat
    DRAG gradient 
    -0.104202

Is SU2_DOT_AD able to compute anything else in this case? Can it compute adjoint variables which are comparable to the continuous adjoint method ones?

Could you please clarify how I should run SU2_DOT_AD in this case?
Thank you!

**Reply by pcarruscag** [reply]
There is only one design variable and thus the of_grad file only has one value.
You can compare it with the result of running SU2_DOT (not ad) on the result of the continuous adjoint

**Reply by frx-wintermute** [reply]
How should I run SU2_DOT?
I tried with:

    $ ln -s restart_adj_cd.dat solution_adj_cd.dat
    $ SU2_DOT inv_adj_NACA0012.cfg

in the directory where I had obtained the continuous adjoint solution, but SU2_DOT exited with the following error:

    Error in "virtual void CPhysicalGeometry::SetBoundSensitivity(CConfig*)": 
    -------------------------------------------------------------------------
    Surface sensitivity not found in file.
    ------------------------------ Error Exit -------------------------------

Where did I go wrong?
